File: python/semantic_kernel/processes/kernel_process/kernel_process_step_state.py
# Copyright (c) Microsoft. All rights reserved.

from typing import Generic, Literal, TypeVar

from pydantic import ConfigDict, Field

from semantic_kernel.kernel_pydantic import KernelBaseModel
from semantic_kernel.utils.feature_stage_decorator import experimental

TState = TypeVar("TState")


@experimental
class KernelProcessStepState(KernelBaseModel, Generic[TState]):
    """The state of a step in a kernel process."""

    type: Literal["KernelProcessStepState"] = Field(default="KernelProcessStepState")  # type: ignore

    name: str
    version: str
    id: str | None = None
    state: TState | None = None

    model_config = ConfigDict(extra="ignore")

File: nodejs/semantic-kernel/processes/kernel-process/kernel-process-step-state.ts
/**
 * The state of a step in a kernel process.
 * @experimental
 */
export class KernelProcessStepState<TState = any> {
  /**
   * The type identifier for this state.
   */
  type: 'KernelProcessStepState' | 'KernelProcessState' = 'KernelProcessStepState'

  /**
   * The name of the step.
   */
  name: string

  /**
   * The version of the step.
   */
  version: string

  /**
   * The optional ID of the step.
   */
  id?: string | null = null

  /**
   * The state data.
   */
  state?: TState | null = null

  constructor(params: { name: string; version: string; id?: string | null; state?: TState | null }) {
    this.name = params.name
    this.version = params.version
    if (params.id !== undefined) {
      this.id = params.id
    }
    if (params.state !== undefined) {
      this.state = params.state
    }
  }
}
